src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+        let client = self.client.as_ref().unwrap();
+
+        loop {
+            trace!(log, "checking spot request status");
+
+            let res = client.describe_spot_instance_requests(req.clone()).sync();
+            if let Err(e) = res {
+                let msg = format!("{}", e);
+                if msg.contains("The spot instance request ID") && msg.contains("does not exist") {
+                    trace!(log, "spot instance requests not yet ready");
+                    continue;
+                } else {
+                    return Err(e)
+                        .context("failed to describe spot instances")
+                        .map_err(|e| e.into());
+                }
+            }
+            let res = res.expect("Err checked above");
+
+            let any_pending = res
+                .spot_instance_requests
+                .as_ref()
+                .expect("describe always returns at least one spot instance")
+                .iter()
+                .map(|sir| {
+                    (
+                        sir,
+                        sir.state
+                            .as_ref()
+                            .expect("spot request did not have state specified"),
+                    )
+                })
+                .any(|(sir, state)| {
+                    if state == "open" || (state == "active" && sir.instance_id.is_none()) {
+                        true
+                    } else {
+                        trace!(log, "spot request ready"; "state" => state, "id" => &sir.spot_instance_request_id);
+                        false
+                    }
+                });
+
+            if !any_pending {
+                // unwraps okay because they are the same as expects above
+                self.instances = res
+                    .spot_instance_requests
+                    .unwrap()
+                    .into_iter()
+                    .filter_map(|sir| {
+                        if sir.state.as_ref().unwrap() == "active" {
+                            // unwrap ok because active implies instance_id.is_some()
+                            // because !any_pending
+                            let instance_id = sir.instance_id.unwrap();
+                            trace!(log, "spot request satisfied"; "iid" => &instance_id);
+
+                            Some((instance_id, (None, self.outstanding_spot_request_ids.remove(&sir.spot_instance_request_id.unwrap()).unwrap())))
+                        } else {
+                            error!(log, "spot request failed: {:?}", &sir.status; "state" => &sir.state.unwrap());
+                            None
+                        }
+                    })
+                    .collect();
+                break;
+            } else {
+                thread::sleep(time::Duration::from_secs(1));
+            }
+
+            if let Some(wait_limit) = max_wait {
+                if start.elapsed() > wait_limit {
```

Comment:
I would probably invert this condition and make it `return Ok(())` early so you avoid the extra indentation.

Reply:
it should be a `continue;` I think?